Create a new hook to enable data decimation

etimberg opened the pull request:
The `beforeElementsUpdate` hook can be used to decimate data. The chart elements will not be created until after this hook has fired ensuring that if decimation occurs, only the needed elements will be created.

This is a simpler take on #8251 that does not need to reorder any passes. Instead, the decimation plugin uses a heuristic to determine the chart width.

## Results

### Test 1 (uPlot Benchmark, 100 runs)

| | Master | PR |
| ---- | ------ | -- |
| Average | 53ms | 20 ms |
| Min | 37ms | 13 ms |
| Max | 175ms | 137ms |
| Variation | 138ms |  124ms |
| Heap | 36425.8 kB | 21191.4 kB |

### Test 2 (10M data points, since run)

| | Master | PR |
| ---- | ------ | -- |
| Render Time | 2954.39 ms | 153.13 ms |

## Decimation Plugin

```javascript
{
	id: 'decimate-data',
	beforeElementUpdate: (chart, args) => {
		// Use an approximation since this is before scales are parsed
		const verticalAxisCount = Object.values(chart.scales).reduce((acc, scale) => acc += scale.axis === 'y' ? 1 : 0, 0);

		// assume ~50px for now. If the axis is actually wider then we have more points than needed
		// TODO: Take display settings into account and compute a more accurate guess for each axis
		const availableWidth = chart.width - (verticalAxisCount * 50);

		chart.data.datasets.forEach((dataset, datasetIndex) => {
			let i, point, x, y, prevX, minIndex, maxIndex, minY, maxY;
			const {data} = dataset;
			const meta = chart.getDatasetMeta(datasetIndex);

			if (meta.type !== 'line') {
				// Only line datasets are supported
				return;
			}

			const xAxis = chart.scales[meta.xAxisID];
			if (xAxis.type !== 'linear' && xAxis.type != 'time') {
				// Only linear interpolation is supported
				return;
			}

			if (!!chart.options.parsing) {
				// Plugin only supports data that does not need parsing
				return;
			}

			const decimated = [];

			const xMin = data[0].x;
			const xMax = data[data.length - 1].x;
			const dx = xMax - xMin;

			for (i = 0; i < data.length; ++i) {
				point = data[i]
				x = (point.x - xMin) / dx * availableWidth;
				y = point.y;
				const truncX = x | 0;

				if (truncX === prevX) {
					// Determine `minY` / `maxY` and `avgX` while we stay within same x-position
					if (y < minY) {
						minY = y;
						minIndex = i;
					} else if (y > maxY) {
						maxY = y;
						maxIndex = i;
					}
				} else {
					// Push up to 4 points, 3 for the last interval and the first point for this interval
					if (minIndex && maxIndex) {
						decimated.push(data[minIndex], data[maxIndex]);
					}
					if (i > 0) {
						// Last point in the previous interval
						decimated.push(data[i - 1]);
					}
					decimated.push(point);
					prevX = truncX;
					minY = maxY = y;
					minIndex = maxIndex = i;
				}
			}

			dataset._data = dataset.data;
			dataset.data = decimated;
		});
	}
};
```

benmccann commented:
I'm not quite sure how to read the results. Does that mean it went from 2086ms to 1487ms + 153ms or just 153ms? Good improvement either way, but if the latter then wow!!

benmccann commented on src/core/core.datasetController.js:
```diff
@@ -399,7 +392,7 @@ export default class DatasetController {
 		], {
 			merger(key, target, source) {
-				if (key !== 'data') {
+				if (key !== 'data' && key.charAt(0) !== '_') {
```
this line could probably use a comment explaining

etimberg replied:
Added a comment. Without it, when the plugin stored the original data onto `dataset._data` it got cloned and slowed everything down. Another option might be to provide some standard storage mechanism for plugins, but this seemed super simple and easy to reason about.

benmccann commented on types/index.esm.d.ts:
```diff
@@ -774,7 +774,15 @@ export interface Plugin<O = {}> extends ExtendedPlugin {
+	 * @param {object} args - The call arguments.
+	 * @param {object} options - The plugin options.
+   */
+  beforeElementUpdate?(chart: Chart, args: {}, options: O): void;
```
we should probably keep `beforeUpdate` and `afterUpdate` next to each other for consistency with the other before/after methods

etimberg replied:
Good call, I moved it below `afterUpdate`.

benmccann reviewed (commented):
I'm assuming if we had this new plugin, then we wouldn't get benefit from the existing line controller decimation anymore? I wonder if we removed the existing line controller decimation and added this general decimation plugin if we'd stay roughly equal in size, but come out ahead on speed and so maybe we should include this plugin in core in that case

etimberg commented:
I've updated that to be clearer. It's two different tests. On the main benchmark it went, on average over 100 runs, from 53ms to 20ms. Rendering 10M points, rendering time dropped from 2954ms to 153ms

etimberg commented:
> 
> 
> I'm assuming if we had this new plugin, then we wouldn't get benefit from the existing line controller decimation anymore? I wonder if we removed the existing line controller decimation and added this general decimation plugin if we'd stay roughly equal in size, but come out ahead on speed and so maybe we should include this plugin in core in that case

@kurkle and I bounced that around too. It might be worth exploring since this also reduces memory overhead. That being said, without the plugin this is quite small.

 **Size Change:** +152 B (0%) 

**Total Size:** 217 kB

| Filename | Size | Change |
| :--- | :---: | :---: |
| `dist/chart.esm.js` | 63.1 kB | +56 B (0%) |
| `dist/chart.js` | 79 kB | +59 B (0%) |
| `dist/chart.min.js` | 57.3 kB | +37 B (0%) |

etimberg commented on src/core/core.controller.js:
```diff
@@ -469,9 +469,14 @@ class Chart {
 		// Make sure dataset controllers are updated and new controllers are reset
 		const newControllers = me.buildOrUpdateControllers();
 
+		me.notifyPlugins('beforeElementUpdate');
```
Not sure what it would be used for, but I have no strong objections

kurkle replied:
Shouldn't it be `beforeElementsUpdate`? 
we have `beforeDatasetsUpdate` and `beforeDatasetUpdate`, so in that context one could think this would fire for each element if its not plural.

etimberg replied:
Good point, I've renamed it to `beforeElementsUpdate`.

leeoniya commented:
+1 for having this behavior in the core and enabled via an opt.

i'm already not a fan of HighCharts needing the additional "boost" module either. at some point i'm going to add `**` callouts for libs that require excessive and precise config to attain the shown bench results.

etimberg commented:
That's some good feedback @leeoniya. I agree that if we can find a way to make it on by default, it'd be great. In my ideal world, I'd try to make it enabled even when we have to parse and sort the data as it would be a good improvement for everyone. With that said, I did look at flot and they have the hook for decimation before parsing.

Some things I'd want to do on the plugin before we consider shipping it:

- [ ] Support more axis types (log, timeseries)
- [ ] Bail out early if `data.length < 4 * width`
- [ ] Testing for heuristics and decide if fancier axis measuring is needed
- [ ] Disable when the primary axis is `y`
- [ ] Find a way to make this work better with editable data. Right now replaced `dataset.data` is disruptive

kurkle commented on src/core/core.datasetController.js:
```diff
@@ -419,9 +415,9 @@ export default class DatasetController {
 		let sorted = true;
 		let i, parsed, cur, prev;
 
-		if (start > 0) {
+		if (start > 0 || (start === 0 && count < data.length)) {
 			sorted = meta._sorted;
-			prev = meta._parsed[start - 1];
+			prev = start > 0 ? meta._parsed[start - 1] : undefined;
 		}
```
```suggestion
		let sorted = start === 0 && count === data.length ? true : meta._sorted;
		let prev = start > 0 && meta._parsed[start - 1];
		let i, cur, parsed;
```

etimberg replied:
Good simplification. With fresh eyes, I thought of some cases where this method could incorrectly detect the sorted status of the data array.

## Incorrectly Mark Data as not Sorted

```
const data = [0, 2, 1, 3, 4];
parse(0, 5); // not sorted

data[1] = 1;
data[2] = 2; // data == [0, 1, 2, 3, 4]

parse(0, 3); // not sorted
```

## Incorrect Mark Data as Sorted

```javascript
const data = [0, 1, 2, 3, 4];
parse(0, 5); // sorted

data[2] = 4; // data == [0, 1, 4, 3, 4]
parse(0, 3); // sorted
```

kurkle replied:
Good points. We should set the `meta._sorted` to false when data is modified (and parsing is enabled).
And add tests for these cases, to make sure it works :)

kurkle reviewed (commented):
I think these new hook(s) could be valuable even if we do the decimation in core.

If doing in core (+1), it might be good to just have that new hook for controllers too, and implement it in the line controller. Does not make sense to have it in 2 levels, so moving from LineElement to LineController works for me.

Note: After this, the `addElements` only adds the dataset element, which is not what you'd except from the function name.

etimberg commented:
> 
> 
> I think these new hook(s) could be valuable even if we do the decimation in core.
> 
> If doing in core (+1), it might be good to just have that new hook for controllers too, and implement it in the line controller. Does not make sense to have it in 2 levels, so moving from LineElement to LineController works for me.
> 
> Note: After this, the `addElements` only adds the dataset element, which is not what you'd except from the function name.

Can you expand a bit on what you're thinking for this? Do you mean having a `beforeElementUpdate` method on the controller that gets called first? Or do you mean a `beforeElementUpdate` hook that the dataset controller fires?

kurkle commented:
> Can you expand a bit on what you're thinking for this? Do you mean having a `beforeElementUpdate` method on the controller that gets called first? Or do you mean a `beforeElementUpdate` hook that the dataset controller fires?

The first one, having a `beforeElementsUpdate` on the controller. I did not walk the code through, so not sure if that makes sense.

etimberg commented:
It might be possible to put it there though not sure if the controller should be modifying the data.

etimberg commented:
I made some progress. I added the plugin into the core (not yet enabled by default). I also added a new feature, `dataset.dataKey` that allows pointing the data stream away from `dataset.data`. My thinking is that users don't want their original array modified, so the decimation plugin stores the decimated data as `dataset._decimated` and then we tell the controller to treat `datset._decimated` as the array to listen to (for updates, etc). That should resolve a lot of the complexity with updates & user interactions.

Work Still To-Do

- [ ] Document decimation plugin options
- [ ] Test all controllers with `dataKey` feature
- [ ] Decide on some initial options for decimation. Allow code to support configurable decimation algorithms in the future

benmccann commented on test/fixtures/controller.bar/data-key.js:
```diff
@@ -0,0 +1,29 @@
+			labels: [0, 1, 3, 4],
+			datasets: [
+				{
+                    data2: [5, 20, 10, 11],
```
spaces -> tabs

benmccann commented on test/fixtures/controller.line/data-key.js:
```diff
@@ -0,0 +1,29 @@
+			labels: [0, 1, 3, 4],
+			datasets: [
+				{
+                    data2: [5, 20, 10, 11],
```
spaces -> tabs

benmccann commented on types/index.esm.d.ts:
```diff
@@ -527,13 +527,14 @@ export class DatasetController<TElement extends Element = Element, TDatasetEleme
 	draw(): void;
 	reset(): void;
-	getDataset(): ChartDataset;
+  getDataset(): ChartDataset;
```
spaces -> tabs

(for every line changed in this file)

etimberg replied:
This entire file appears to be spaces.

benmccann replied:
It looks really inconsistent. The first 379 lines are spaces, the next 900 are tabs, and it keeps switching back and forth

We should probably convert the whole thing to tabs at some point though that seems outside the scope of this PR

Still it might be a good idea to have the new lines you're adding here be tabs because the lines before and after them are tabs and so things don't align properly because of the mismatch

etimberg replied:
Ok, looks like there's a section of tabs in the middle. I had looked at the bottom and it's spaces. Let's do a separate PR to fix that since it will complicate the diff here

benmccann replied:
https://github.com/chartjs/Chart.js/pull/8263 has been merged to standardize the file on tabs, so the lines be added in this PR should be switched to tabs now too

benmccann commented:
`dataKey` seems a little risky to me because every plugin, etc. would have to understand to get data from `dataKey` instead of `data` which would add complication. I'd probably just stick the original data on another field called `_originalData` and replace `data` with the decimated data. The original array isn't be modified (just moved), so the user can easily maintain a reference to it outside the chart to maintain access to it without having to know about the `_originalData` field.

kurkle commented:
We could create a getter for the data that uses `dataKey`. Would need to rename the original data property then to something else and default dataKey to that.

etimberg commented:
I thought about the getter solution too. The reason I didn't implement it was what happens if someone did something like #7340 and passed the data in a class instance

benmccann commented:
I'm not sure I understand how the class instance relates to having a getter. They seem orthogonal to me, but I might be missing something

etimberg commented:
I meant if the user passes in a class where `data` is already a getter, we may not be able to redefine it. This is also likely a problem for frozen data objects (hence why we moved the metas to `chart._metasets`). Perhaps we can put the data key there, or as you said, just modify the data array and tell the user that that's how the plugin works.

@kurkle @simonbrunel any thoughts?

kurkle commented:
The use passed data object can be stored in the config. we can build the `chart.data.datasets[i]` and have getter for that objects `data` property.
so `chart.config.data.datasets[0].data` would be the original data object. when decimated we lose any additional properties or prototype a data point has. What are we thinking differently?

kurkle commented:
Some clarification to what I was suggesting:

currently:
```js
chart {
  get data { return this.config.data; }
  set data(data) { this.config.data = data; }
}
```

proposal:
```js
chart {
  constructor() {
    this.data = { labels: [], datasets: [] };
  }
  set data(data) { 
    this.config.data = data;
    // update('data'), or continue requiring user to call update()
  }
  buildOrUpdateControllers() {
    const datasets = this.data.datasets;
    const configDatasets = me.config.data.datasets;
    for (let i = 0, ilen = configDatasets .length; i < ilen; i++) {
      // simplified, update logic should remain
      datasets[i] = new ControllerClass(me, i, configDatasets[i]);
    }
  }

}
```

So `chart.data.datasets` array would contain the controllers, but setting `chart.data` would equal to setting `chart.config.data` (for compatibility, not sure if its needed).

Could also combine controller and its meta for reduced complexity and overhead, but I'm not sure if it would be good in all aspects.

benmccann commented:
@kurkle's idea seems fairly reasonable to me

etimberg commented:
I'm trying to think more about this. It could be very surprising if  `chart.data.datasets` was the controllers and many people may be using `chart.data` and assuming it is the data object they passed in. In v3 we did a lot of work to stop modify the `data` object passed by the user and this feels like a regression backwards.

Maybe I should just update the plugin to modify the dataset and keep the changes minimal force it to be opt-in. Something like this. When decimating in future updates, the plugin would need to look for `_data` first since that indicates whether or not the dataset has been mutated.

```javascript

dataset._data = dataset.data;
dataset._decimated = dataset.decimated;

Object.defineProperty(dataset, 'data', {
  configurable: true,
  enumerable: true,
  writable: true,
  get: function() {
    return this._decimated;
  },
  set: function(data) {
    this._data = data;
  }
});
```

We'd probably want something to clean this up on chart delete and put the object back to what it was.

benmccann commented:
> I'm trying to think more about this. It could be very surprising if chart.data.datasets was the controllers and many people may be using chart.data and assuming it is the data object they passed in. In v3 we did a lot of work to stop modify the data object passed by the user and this feels like a regression backwards.

I agree with all of this.

> Maybe I should just update the plugin to modify the dataset and keep the changes minimal force it to be opt-in.

I still think we might be able to do better than this

Here's another idea - what if we made `meta.data` hold the decimated data and `chart.data.datasets.data` could remain the original data? We'd have to be a little careful of which one we're using in the scriptable context and plugins, but hopefully it should mostly be pretty straightforward

etimberg commented:
I thought about something like that. `meta.data` is currently the elements, but `meta.parsed` is the parsed data (that would be decimated). The big downside that I thought of is that most of our examples would break subtly when decimating. We show the callbacks for the tooltips using `dataset.data[index]` and without decimation the index will match up, but as soon as there is any decimation, the indexes will be off and user tooltips could show the wrong data. I can see this leading to a lot of issues as users migrate over to v3.

benmccann commented:
Callbacks for tooltips are much simpler if they just use `dataPoint` or `formattedValue` ([docs](https://www.chartjs.org/docs/next/configuration/tooltip#tooltip-item-context)). Perhaps we should remove `dataset` from the context to avoid the possibility of users subtly tripping up as you mentioned and to encourage them to write it in the simpler way. I took a look at this and the one thing we'd need to change is that `dataPoint` is only the parsed value. We should probably add a `rawPoint` as well

etimberg commented:
Agreed that callbacks should use `dataPoint` or `formattedValue`, but if someone just skims the v3 docs and migrates, it will more or less work in simple cases which is my concern. @kurkle @simonbrunel any thoughts?

benmccann commented:
Not sure what you mean by "more or less" - maybe you just meant "more"? Removing `dataset` would make it obvious for people, but if we don't want to do that it's rare that a chart has enough data to trigger decimation so maybe not that big a risk to leave it. Or we could also leave it and deprecate it - make it a get method that also prints a warning

etimberg commented:
My bad, that was an incomplete sentence. I was, as you said, trying to get across that it mostly works as intended unless decimation happens. I think that might be OK since decimation will be opt-in, but will let everyone else chime in before I implement.

etimberg commented:
Some changes.

1. Rebased against the latest from master
2. Removed the `dataKey` implementation and replaced using the descriptor idea above (still needs some testing to see how updates behave)

kurkle commented on src/plugins/plugin.decimation.js:
```diff
@@ -0,0 +1,124 @@
+
+	beforeElementsUpdate: (chart, args, options) => {
+		// Decimation is only supported for lines that have an X indexAxis
+		if (!options.enabled || chart.options.indexAxis === 'y') {
```
The `indexAxis` option is also configurable per dataset, so for completeness it should be checked in the loop below.

kurkle commented on src/plugins/plugin.decimation.js:
```diff
@@ -0,0 +1,124 @@
+		// Use an approximation since this is before scales are parsed
+		const verticalAxisCount = Object.values(chart.scales).reduce((acc, scale) => acc + scale.axis === 'y' ? 1 : 0, 0);
+
+		// assume ~50px for now. If the axis is actually wider then we have more points than needed
```
Should we actually just assume the whole width here, because the axes can be hidden? Should still be quite efficient.
It might look bad though.

etimberg replied:
> 
> 
> This looks like a clean solution.
> I wrote 2 comments to the available width calculation, that go in different directions regarding to the accuracy. I'm not suggesting either of those should be made here, only sharing thoughts.

Thanks @kurkle! i'll take a look and get back to you. I'm sort of leaning to the simple solution of assuming the full width since it errs on the side of adding more points.

etimberg replied:
I changed to simplify and just assume we have `chart.width` available to use. It'll have a few extra points, but not too many.

kurkle commented on src/plugins/plugin.decimation.js:
```diff
@@ -0,0 +1,124 @@
+
+		// assume ~50px for now. If the axis is actually wider then we have more points than needed
+		// TODO: Take display settings into account and compute a more accurate guess for each axis
+		const availableWidth = chart.width - (verticalAxisCount * 50);
```
This could probably be done by  adding a method to scale API for the size calculation (that would be used in layout too).

kurkle reviewed (commented):
This looks like a clean solution.
I wrote 2 comments to the available width calculation, that go in different directions regarding to the accuracy. I'm not suggesting either of those should be made here, only sharing thoughts.

benmccann commented:
Should we go ahead and remove the existing line decimation or did you prefer to leave that as a follow up for another PR?

etimberg commented:
I think it'd be good to leave that for another PR

kurkle commented:
> Should we go ahead and remove the existing line decimation or did you prefer to leave that as a follow up for another PR?

I'm not sure we should remove it, because the late decimation works in wider range of cases (no scale type limitation and data can be immutable)

kurkle commented on src/plugins/plugin.decimation.js:
```diff
@@ -0,0 +1,124 @@
+				// First time we are seeing this dataset
+				// We override the 'data' property with a setter that stores the
+				// raw data in _data, but reads the decimated data from _decimated
+				// TODO: Undo this on chart destruction
```
Something like this?
```js
destroy(chart, args, options) {
  chart.data.datasets.forEach((dataset, datasetIndex) => {
    if (dataset._decimated) {
      const data = dataset._data;
      delete dataset._decimated;
      delete dataset._data;
      Object.defineProperty(dataset, 'data', { value: data });
    }
  });
}

etimberg replied:
Yup, I used that. Thanks!

kurkle commented on src/plugins/plugin.decimation.js:
```diff
@@ -0,0 +1,124 @@
+				return;
+			}
+
+			// Point the chart to the decimated data
```
should the `defineProperty` be moved down here, after the bailouts?

etimberg replied:
It should, I've updated accordingly

kurkle reviewed (commented):
Couple of comments on the plugin, everything else looks ready to me.

benmccann commented on src/plugins/plugin.decimation.js:
```diff
@@ -0,0 +1,124 @@
+			}
+
+			const xAxis = chart.scales[meta.xAxisID];
+			if (xAxis.type !== 'linear' && xAxis.type !== 'time') {
```
I think `timeseries` should work as well. And probably `category` too. I can't imagine how we'd plot those with more points than pixels. The only thing I can think is that we might want to skip the plugin if `options.ticks.major.enabled` because in that case we might want to let autoskip do it's job to pick the important points to put on the axis

etimberg replied:
See @kurkle's comment below for why this doesn't work at the moment. re major ticks, I don't think those settings affect the drawing of the line.

benmccann commented:
> I'm not sure we should remove it, because the late decimation works in wider range of cases (no scale type limitation and data can be immutable)

I can't remember the details of the existing one. Is its algorithm much different? Why would this plugin need a scale type limitation while the existing one doesn't?  I wonder if we should have an option to force the plugin to apply so that someone can use it with a custom scale.

I still think it'd be nice to make this plugin work with immutable data :smile:

In either case leaving the existing one in for this PR makes sense to me

kurkle commented:
@benmccann the difference between plugin and LineController decimation is, the plugin uses the data values and assumes linear distance between values, while the one in controller uses the actual coordinates from elements. That's why this plugin does not work with timeseries or category, those are not linear.

kurkle commented:
@etimberg 
https://github.com/leeoniya/uPlot/pull/455#issuecomment-778391754

etimberg commented:
I think we might need to account for that `(0.5, 0.5)` translation that is applied
